nemo/collections/nlp/models/duplex_text_normalization/duplex_tagger.py

```diff
@@ -0,0 +1,315 @@
+        super().__init__(cfg=cfg, trainer=trainer)
+        self.num_labels = len(constants.ALL_TAG_LABELS)
+        self.model = AutoModelForTokenClassification.from_pretrained(cfg.transformer,
+                                                                    num_labels=self.num_labels)
+
+        # Loss Functions
+        self.loss_fct = nn.CrossEntropyLoss(ignore_index=constants.LABEL_PAD_TOKEN_ID)
+
+        # Device
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
+        self.to(device)
+
+        # For validation
+        self.val_all_preds, self.val_all_targets = [], []
+
+    # Training
+    def training_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the training loop with the data from the training dataloader
+        passed in as `batch`.
+        """
+        self.train()
+        num_labels = self.num_labels
+
+        # Apply Transformer
+        input_ids = batch['input_ids'].to(self.device)
+        input_masks = batch['attention_mask'].to(self.device)
+        tag_logits = self.model(input_ids, input_masks).logits
+
+        # Loss
+        tag_labels = batch['labels'].view(-1).to(self.device)
+        train_loss = self.loss_fct(tag_logits.view(-1, num_labels), tag_labels)
+
+        lr = self._optimizer.param_groups[0]['lr']
+        self.log('train_loss', train_loss)
+        self.log('lr', lr, prog_bar=True)
+        return {
+            'loss': train_loss,
+            'lr': lr
+        }
+
+    # Validation and Testing
+    def validation_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the validation loop with the data from the validation dataloader
+        passed in as `batch`.
+        """
+        self.eval()
+        num_labels = self.num_labels
+
+        # Apply Transformer
+        input_ids = batch['input_ids'].to(self.device)
+        input_masks = batch['attention_mask'].to(self.device)
+        tag_logits = self.model(input_ids, input_masks).logits
+        tag_preds = torch.argmax(tag_logits, dim=2)
+
+        # Loss
+        tag_labels = batch['labels'].to(self.device)
+        val_loss = self.loss_fct(tag_logits.view(-1, num_labels),
+                                 tag_labels.view(-1))
+
+        # Extract batch_predictions and batch_labels
+        predictions, labels = tag_preds.tolist(), tag_labels.tolist()
+        final_predictions = [
+            [constants.ALL_TAG_LABELS[p] for (p, l) in zip(prediction, label) \
+             if l != constants.LABEL_PAD_TOKEN_ID]
+            for prediction, label in zip(predictions, labels)
+        ]
+        final_labels = [
+            [constants.ALL_TAG_LABELS[l] for (p, l) in zip(prediction, label) \
+             if l != constants.LABEL_PAD_TOKEN_ID]
+            for prediction, label in zip(predictions, labels)
+        ]
+        self.val_all_preds.extend(final_predictions)
+        self.val_all_targets.extend(final_labels)
```

Comment:
you can accumulate outputs across batches by returning val_all_preds and val_all_targets, see https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/question_answering/qa_model.py#L114

Reply:
Yes. I modified that.